packages/eslint-plugin/docs/rules/no-unnecessary-type-parameters.md

```diff
@@ -0,0 +1,104 @@
+---
+description: 'Disallow type parameters that only appear once'
+---
+
+> 🛑 This file is source code, not the primary documentation location! 🛑
+>
+> See **https://typescript-eslint.io/rules/no-unnecessary-type-parameters** for documentation.
+
+This rule forbids type parameters that only appear once in a function, method, or class definition.
+
+Type parameters relate two types. If a generic type parameter only appears once, then it is not relating anything. It can usually be replaced with explicit types such as `unknown`.
+
+At best unnecessary type parameters make code harder to read. At worst they can be used to disguise unsafe type assertions.
+
+## Options
+
+None.
+
+## Examples
+
+<!--tabs-->
+
+### ❌ Incorrect
+
+```ts
+function third<A, B, C>(a: A, b: B, c: C): C {
+  return c;
+}
+
+function parseYAML<T>(input: string): T {
+  // ...
+}
+
+function printProperty<T, K extends keyof T>(obj: T, key: K) {
+  console.log(obj[key]);
+}
+
+class Joiner<T extends string | number> {
+  join(els: T[]) {
+    return els.map(el => '' + el).join(',');
+  }
+}
+```
+
+### ✅ Correct
+
+```ts
+function third<A, B, C>(a: unknown, b: unknown, c: C): C {
+  return c;
+}
+
+function parseYAML(input: string): unknown {
+  // ...
+}
+
+function printProperty<T>(obj: T, key: keyof T) {
+  console.log(obj[key]);
+}
+
+class Joiner {
+  join(els: (string | number)[]) {
+    return els.map(el => '' + el).join(',');
+  }
+}
+
+// T appears twice: in the type of arg and as the return type
+function identity<T>(arg: T): T {
+  return arg;
+}
+
+// T appears twice: "keyof T" and in the inferred return type (T[K]).
+// K appears twice: "key: K" and in the inferred return type (T[K]).
+function getProperty<T, K extends keyof T>(obj: T, key: K) {
+  return obj[key];
+}
+```
+
+## When Not To Use It
+
+This rule will prevent using type parameters to test types, for example:
+
+```ts
+function assertType<T>(arg: T) {}
+
+assertType<number>(123);
+assertType<number>('abc');
+//                 ~~~~~
+// Argument of type 'string' is not assignable to parameter of type 'number'.
+```
+
+If you're using this pattern then you'll want to disable this rule.
+
+If you're using generic type parameters to implement phantom types then you'll also run afoul of this rule.
```

Comment:
[Docs] I don't know what phantom types are, and neither will many readers of this. This should either define it or be omitted. A quick definition and link to further reading would be sufficient.